# <span style="color:red">TASKS</span>

* update data types
* Do I split new_applicants data 70% and 30%... or do I use the training data given to me? based on titanic video I might split my training data, to train the model, then use unknown dataset to old_applicants the model... 3 vids can help... 
    * [data school](https://www.youtube.com/watch?v=RlQuVL6-qe8&index=4&list=PL5-da3qGB5ICeMbQuqbbCOQWcS6OYBr5A): 
    * [packt](https://www.youtube.com/watch?v=Eqv98w1ukZk): splits data set as projedct instructions describe
    * [titanic](https://www.youtube.com/watch?v=VCJdg7YBbAQ): describes creating dummies from dataset
    * [iris google dev]([iris google developers](https://www.youtube.com/watch?v=tNa99PG8hR8&list=PLOU2XLYxmsIIuiBfYad6rFYQU_jL2ryal&index=2)

* Create dummies on fields with multiple numeric/text values (e.g. sex, ticket class)
* Correlation & P-values, select predictors
    * watch udacity videos and other videos if needed
    
* Does target variable also need to be converted to numeric, or is that only an Altryx issue???
    * if so create new column np.where([old_applicants["Credit-Application-Result"] == 'Creditworthy', 1, 0)

# RESOURCES

* ## [project details](https://classroom.udacity.com/nanodegrees/nd008t/parts/54e2103b-5630-497a-bc37-c05d45cda549/modules/12c50013-ad11-4980-88a0-13d474cda9b8/lessons/ee56f886-070e-40ca-bfc8-cab2c5cdde7d/concepts/ac7bf309-3e98-4a2f-ad4e-555905d7cd07)
* ## [rubric](https://review.udacity.com/#!/rubrics/265/view)
* ## [main project folder](https://udacity-travismgillespie.notebooks.azure.com/j/tree)
* ## variance
    * ### [use R-squared to describe variance](https://classroom.udacity.com/nanodegrees/nd008t/parts/9cd2f005-f9b1-4953-ba02-ad65805b2a4a/modules/fc4c1ffe-01f6-4c11-b38b-646ecf1d3c59/lessons/576552b9-a1b4-45be-9428-572331a4e186/concepts/82bfe969-921b-4175-adb5-797e84ba6107): also look into [project 2](https://udacity-travismgillespie.notebooks.azure.com/j/notebooks/project2/Data%20Wrangling.ipynb) diamond catalog to see if I calculated correlation p and r-squared at same time... may have used a corr-matrix
* ## feature selection
    * ### [Packt: select features in machine learning](https://hub.packtpub.com/selecting-statistical-based-features-in-machine-learning-application/)
    * ### [use p-value & correlation to select predictor variable](https://classroom.udacity.com/nanodegrees/nd008t/parts/9cd2f005-f9b1-4953-ba02-ad65805b2a4a/modules/fc4c1ffe-01f6-4c11-b38b-646ecf1d3c59/lessons/576552b9-a1b4-45be-9428-572331a4e186/concepts/82bfe969-921b-4175-adb5-797e84ba6107)
    * ### [p values don’t help in feature selection](https://www.quora.com/How-did-you-use-p-value-for-feature-selection)
    * ### [Feature selection - univariate selection](https://blog.datadive.net/selecting-good-features-part-i-univariate-selection/)
    * ### [predictor variables](https://classroom.udacity.com/nanodegrees/nd008t/parts/9cd2f005-f9b1-4953-ba02-ad65805b2a4a/modules/fc4c1ffe-01f6-4c11-b38b-646ecf1d3c59/lessons/576552b9-a1b4-45be-9428-572331a4e186/concepts/631d190c-8626-4dd7-92df-f5bd96913c48): udacity
    * ### [explanatory variables (aka predictor variables)](https://classroom.udacity.com/nanodegrees/nd008t/parts/9cd2f005-f9b1-4953-ba02-ad65805b2a4a/modules/fc4c1ffe-01f6-4c11-b38b-646ecf1d3c59/lessons/576552b9-a1b4-45be-9428-572331a4e186/concepts/631d190c-8626-4dd7-92df-f5bd96913c48): udacity
    * ### [feature selection ex](https://www.youtube.com/watch?v=hhmsHY2vWY4)
    * ### [feature selection correlation and p-value](https://towardsdatascience.com/feature-selection-correlation-and-p-value-da8921bfb3cf)
* ## coefficients
    * ### [calculate coefficients](https://classroom.udacity.com/nanodegrees/nd008t/parts/9cd2f005-f9b1-4953-ba02-ad65805b2a4a/modules/fc4c1ffe-01f6-4c11-b38b-646ecf1d3c59/lessons/576552b9-a1b4-45be-9428-572331a4e186/concepts/631d190c-8626-4dd7-92df-f5bd96913c48): watch udacity video
* ## linear regression
    * ### [packt ex](https://www.youtube.com/watch?v=Eqv98w1ukZk): good example of seeds and splitting test set
* ## logistic regression
    * ### [StatQuest](https://www.youtube.com/watch?v=yIYKR4sgzI8)
    * ### [data school ex](https://www.youtube.com/watch?v=RlQuVL6-qe8&index=4&list=PL5-da3qGB5ICeMbQuqbbCOQWcS6OYBr5A)
    * ### [Titanic ex](https://www.youtube.com/watch?v=VCJdg7YBbAQ)
    * ### [logistic regression ex](https://www.youtube.com/watch?v=1nWFHa6K23w)
* ## decision tree
    * ### [iris google developers](https://www.youtube.com/watch?v=tNa99PG8hR8&list=PLOU2XLYxmsIIuiBfYad6rFYQU_jL2ryal&index=2)
* ## boosted tree (aka Gradient boosting)
    * ### [documentation: scikit learn](https://scikit-learn.org/stable/modules/ensemble.html#gradient-tree-boosting)
    * ### [gradient boosting ex](https://www.youtube.com/watch?v=gIJHI5jSChM)
    * ### [gradient boosting ex](https://www.youtube.com/watch?v=sRktKszFmSk)
    * ### [gradient boosting ex](https://www.youtube.com/watch?v=X3Wbfb4M33w)
    * ### [gradient boosting ex](https://www.youtube.com/watch?v=2WfKzyi__l4)
    * ### [ada boosting ex](https://www.youtube.com/watch?v=GM3CDQfQ4sw)
    * ### [additional boosting reading](https://en.wikipedia.org/wiki/Gradient_boosting)
* ## forest model (aka random forest model)
    * ### [documentation: scikit learn](https://scikit-learn.org/stable/modules/ensemble.html#random-forests)
    * ### [forest model udacity](https://classroom.udacity.com/nanodegrees/nd008t/parts/54e2103b-5630-497a-bc37-c05d45cda549/modules/12c50013-ad11-4980-88a0-13d474cda9b8/lessons/59e0ee88-1a1e-459d-a79e-53ef74c13e77/concepts/65b13e59-9be3-45f8-a6f3-196369dc3c3d)
    * ### [random forest example](https://www.youtube.com/watch?v=D_2LkhMJcfY)
    
* ## backward elemenation
    * ### [looking up details on SL = 0.05](https://www.kaggle.com/umeshsati54/backward-elimination)

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# not sure which train_test_split to use... are they different?
from sklearn.model_selection import train_test_split
from sklearn.cross_validation import train_test_split

from sklearn import tree

from sklearn.linear_model import LogisticRegression

from sklearn.ensemble import GradientBoostingClassifier

from sklearn.ensemble import RandomForestClassifier



/home/nbuser/anaconda3_501/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


## Train Data Set

In [2]:
#training set contains one additional column Credit Application Result
old_applicants_original = pd.read_excel('./data/credit-data-training.xlsx')
old_applicants = old_applicants_original.copy()

In [3]:
# given provided training set has been cleaned shouldn't need to be modified
# consider dropping age where nill
# old_applicants["Age-years"].unique()
# old_applicants["Duration-in-Current-address"].unique()
old_applicants.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 20 columns):
Credit-Application-Result            500 non-null object
Account-Balance                      500 non-null object
Duration-of-Credit-Month             500 non-null int64
Payment-Status-of-Previous-Credit    500 non-null object
Purpose                              500 non-null object
Credit-Amount                        500 non-null int64
Value-Savings-Stocks                 500 non-null object
Length-of-current-employment         500 non-null object
Instalment-per-cent                  500 non-null int64
Guarantors                           500 non-null object
Duration-in-Current-address          156 non-null float64
Most-valuable-available-asset        500 non-null int64
Age-years                            488 non-null float64
Concurrent-Credits                   500 non-null object
Type-of-apartment                    500 non-null int64
No-of-Credits-at-this-Bank           500 no

## new_applicants Data Set

In [4]:
#new_applicants set missing one column... Credit Application Result is the result you are tryting to predict
new_applicants = pd.read_excel('./data/customers-to-score.xlsx')
# new_applicants.sort_values('Purpose', inplace=True)
new_applicants.head()

Account-Balance  Duration-of-Credit-Month Payment-Status-of-Previous-Credit  \
0      No Account                         9        No Problems (in this bank)   
1      No Account                        12        No Problems (in this bank)   
2      No Account                        24                           Paid Up   
3      No Account                        11        No Problems (in this bank)   
4      No Account                         6        No Problems (in this bank)   

        Purpose  Credit-Amount Value-Savings-Stocks  \
0  Home Related           2799                 None   
1  Home Related           2122                 None   
2  Home Related           3758           £100-£1000   
3  Home Related           3905                 None   
4  Home Related           1957                 None   

  Length-of-current-employment  Instalment-per-cent Guarantors  \
0                        < 1yr                    2       None   
1                        < 1yr                    3       None   
2                        < 1yr                    1       None   
3                        < 1yr                    2       None   
4                      1-4 yrs                    1       None   

   Duration-in-Current-address  Most-valuable-available-asset  Age-years  \
0                            2                              1         36   
1                            2                              1         39   
2                            4                              4         23   
3                            2                              1         36   
4                            4                              3         31   

  Concurrent-Credits  Type-of-apartment No-of-Credits-at-this-Bank  \
0  Other Banks/Depts                  1                More than 1   
1  Other Banks/Depts                  1                More than 1   
2  Other Banks/Depts                  1                          1   
3  Other Banks/Depts                  1                More than 1   
4  Other Banks/Depts                  2                          1   

   Occupation  No-of-dependents  Telephone  Foreign-Worker  
0           1                 2          1               1  
1           1                 2          1               2  
2           1                 1          1               1  
3           1                 2          1               1  
4           1                 1          1               1

new_applicants data set has no null values.

In [5]:
new_applicants.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 19 columns):
Account-Balance                      500 non-null object
Duration-of-Credit-Month             500 non-null int64
Payment-Status-of-Previous-Credit    500 non-null object
Purpose                              500 non-null object
Credit-Amount                        500 non-null int64
Value-Savings-Stocks                 500 non-null object
Length-of-current-employment         500 non-null object
Instalment-per-cent                  500 non-null int64
Guarantors                           500 non-null object
Duration-in-Current-address          500 non-null int64
Most-valuable-available-asset        500 non-null int64
Age-years                            500 non-null int64
Concurrent-Credits                   500 non-null object
Type-of-apartment                    500 non-null int64
No-of-Credits-at-this-Bank           500 non-null object
Occupation                           500 non-nu

# CLEAN DATA

## Clean old_applicants Data

In [6]:
old_applicants.isnull().sum()

Credit-Application-Result              0
Account-Balance                        0
Duration-of-Credit-Month               0
Payment-Status-of-Previous-Credit      0
Purpose                                0
Credit-Amount                          0
Value-Savings-Stocks                   0
Length-of-current-employment           0
Instalment-per-cent                    0
Guarantors                             0
Duration-in-Current-address          344
Most-valuable-available-asset          0
Age-years                             12
Concurrent-Credits                     0
Type-of-apartment                      0
No-of-Credits-at-this-Bank             0
Occupation                             0
No-of-dependents                       0
Telephone                              0
Foreign-Worker                         0
dtype: int64

_Duration-in-Current-address_ and _Age-years_ requires some cleanup:

* _Duration-in-Current-address_: immediately removing this attribute since ≈68% of the field is null

* _Age-years_: will handle this cleanup after splitting old_applications into training and testing set. Otherwise I am violating a core tenent in the machine learning procedure. I cannot assume I know the median value of the entire data set to make predictions. I can only assume I know the median value of the training set. In other words I don't want to fit information from my testing set into my training set.

Note:
1. Use the median age of _Age-years_ column to impute missing values.
* Dropping _Duration-in-Current-address_ becasue more than half field is null... ≈ 69%

In [7]:
# calculate percent of null values in Duration-in-Current-address
old_applicants["Duration-in-Current-address"].isnull().sum()/old_applicants["Duration-in-Current-address"].shape[0]

0.688

In [8]:
# Dropping Duration-in-Current-address becasue more than half field is null
old_applicants.drop('Duration-in-Current-address', axis=1, inplace=True)
# there are no more no values in the old_applicants dataset
old_applicants.isnull().sum()

Credit-Application-Result             0
Account-Balance                       0
Duration-of-Credit-Month              0
Payment-Status-of-Previous-Credit     0
Purpose                               0
Credit-Amount                         0
Value-Savings-Stocks                  0
Length-of-current-employment          0
Instalment-per-cent                   0
Guarantors                            0
Most-valuable-available-asset         0
Age-years                            12
Concurrent-Credits                    0
Type-of-apartment                     0
No-of-Credits-at-this-Bank            0
Occupation                            0
No-of-dependents                      0
Telephone                             0
Foreign-Worker                        0
dtype: int64

## Clean new_applicants Data

In [9]:
new_applicants.isnull().sum()

Account-Balance                      0
Duration-of-Credit-Month             0
Payment-Status-of-Previous-Credit    0
Purpose                              0
Credit-Amount                        0
Value-Savings-Stocks                 0
Length-of-current-employment         0
Instalment-per-cent                  0
Guarantors                           0
Duration-in-Current-address          0
Most-valuable-available-asset        0
Age-years                            0
Concurrent-Credits                   0
Type-of-apartment                    0
No-of-Credits-at-this-Bank           0
Occupation                           0
No-of-dependents                     0
Telephone                            0
Foreign-Worker                       0
dtype: int64

As indicated in the code above, and using new_applicants.info() earlier, the new_applicants data set has no null values. Will leave this data set alone for now since there is no immediate clean up needed. However, might be some columns that are removed later from the dataset for other reasons.

# Get Dummy Variables

The string values need to be converted to numerical values

In [10]:
old_applicants.head(10)

Credit-Application-Result Account-Balance  Duration-of-Credit-Month  \
0              Creditworthy    Some Balance                         4   
1              Creditworthy    Some Balance                         4   
2              Creditworthy    Some Balance                         4   
3              Creditworthy    Some Balance                         4   
4              Creditworthy      No Account                         6   
5              Creditworthy    Some Balance                         6   
6          Non-Creditworthy      No Account                         6   
7              Creditworthy      No Account                         6   
8              Creditworthy      No Account                         6   
9              Creditworthy    Some Balance                         6   

  Payment-Status-of-Previous-Credit       Purpose  Credit-Amount  \
0                           Paid Up         Other           1494   
1                           Paid Up  Home Related           1494   
2        No Problems (in this bank)  Home Related           1544   
3        No Problems (in this bank)  Home Related           3380   
4                           Paid Up  Home Related            343   
5        No Problems (in this bank)  Home Related            362   
6                     Some Problems  Home Related            433   
7                           Paid Up  Home Related            454   
8                           Paid Up  Home Related            484   
9                           Paid Up  Home Related            660   

  Value-Savings-Stocks Length-of-current-employment  Instalment-per-cent  \
0           £100-£1000                        < 1yr                    1   
1           £100-£1000                        < 1yr                    1   
2                 None                      1-4 yrs                    2   
3                 None                      1-4 yrs                    1   
4                 None                        < 1yr                    4   
5               < £100                        < 1yr                    4   
6           £100-£1000                        < 1yr                    4   
7                 None                        < 1yr                    3   
8                 None                      1-4 yrs                    3   
9           £100-£1000                      1-4 yrs                    2   

  Guarantors  Most-valuable-available-asset  Age-years Concurrent-Credits  \
0       None                              1        NaN  Other Banks/Depts   
1       None                              1       29.0  Other Banks/Depts   
2       None                              1       42.0  Other Banks/Depts   
3       None                              1       37.0  Other Banks/Depts   
4       None                              1       27.0  Other Banks/Depts   
5       None                              3       52.0  Other Banks/Depts   
6       None                              2       24.0  Other Banks/Depts   
7       None                              2       22.0  Other Banks/Depts   
8        Yes                              1       28.0  Other Banks/Depts   
9       None                              1       23.0  Other Banks/Depts   

   Type-of-apartment No-of-Credits-at-this-Bank  Occupation  No-of-dependents  \
0                  2                          1           1                 2   
1                  2                          1           1                 2   
2                  2                More than 1           1                 2   
3                  2                          1           1                 2   
4                  2                          1           1                 1   
5                  2                More than 1           1                 1   
6                  1                          1           1                 2   
7                  2                          1           1                 1   
8                  2               

In [11]:
pd.options.display.max_columns = 30

In [12]:
# np.where(old_applicants["Credit-Application-Result"] == "Creditworthy", 1, 0)

In [13]:
# old_applicants_dummies = 
old_applicants_dummies = pd.get_dummies(old_applicants, columns = ['Credit-Application-Result', 'Account-Balance', 'Payment-Status-of-Previous-Credit', 'Purpose', 'Value-Savings-Stocks', 'Length-of-current-employment', 'Guarantors', 'Concurrent-Credits', 'No-of-Credits-at-this-Bank'], drop_first=True)


In [14]:
old_applicants_dummies.head()


Duration-of-Credit-Month  Credit-Amount  Instalment-per-cent  \
0                         4           1494                    1   
1                         4           1494                    1   
2                         4           1544                    2   
3                         4           3380                    1   
4                         6            343                    4   

   Most-valuable-available-asset  Age-years  Type-of-apartment  Occupation  \
0                              1        NaN                  2           1   
1                              1       29.0                  2           1   
2                              1       42.0                  2           1   
3                              1       37.0                  2           1   
4                              1       27.0                  2           1   

   No-of-dependents  Telephone  Foreign-Worker  \
0                 2          1               2   
1                 2          1               2   
2                 2          1               1   
3                 2          1               1   
4                 1          1               1   

   Credit-Application-Result_Non-Creditworthy  Account-Balance_Some Balance  \
0                                           0                             1   
1                                           0                             1   
2                                           0                             1   
3                                           0                             1   
4                                           0                             0   

   Payment-Status-of-Previous-Credit_Paid Up  \
0                                          1   
1                                          1   
2                                          0   
3                                          0   
4                                          1   

   Payment-Status-of-Previous-Credit_Some Problems  Purpose_New car  \
0                                                0                0   
1                                                0                0   
2                                                0                0   
3                                                0                0   
4                                                0                0   

   Purpose_Other  Purpose_Used car  Value-Savings-Stocks_None  \
0              1                 0                          0   
1              0                 0                          0   
2              0                 0                          1   
3              0                 0                          1   
4              0                 0                          1   

   Value-Savings-Stocks_£100-£1000  Length-of-current-employment_4-7 yrs  \
0                                1                                     0   
1                                1                                     0   
2                                0                                     0   
3                                0                                     0   
4                                0                                     0   

   Length-of-current-employment_< 1yr  Guarantors_Yes  \
0                                   1               0   
1                                   1               0   
2                                   0               0   
3                                   0               0   
4                                   1               0   

   No-of-Credits-at-this-Bank_More than 1  
0                                       0  
1                                       0  
2                                       1  
3                                       0  
4                                       0

In [15]:
old_applicants_dummies.columns


Index(['Duration-of-Credit-Month', 'Credit-Amount', 'Instalment-per-cent',
       'Most-valuable-available-asset', 'Age-years', 'Type-of-apartment',
       'Occupation', 'No-of-dependents', 'Telephone', 'Foreign-Worker',
       'Credit-Application-Result_Non-Creditworthy',
       'Account-Balance_Some Balance',
       'Payment-Status-of-Previous-Credit_Paid Up',
       'Payment-Status-of-Previous-Credit_Some Problems', 'Purpose_New car',
       'Purpose_Other', 'Purpose_Used car', 'Value-Savings-Stocks_None',
       'Value-Savings-Stocks_£100-£1000',
       'Length-of-current-employment_4-7 yrs',
       'Length-of-current-employment_< 1yr', 'Guarantors_Yes',
       'No-of-Credits-at-this-Bank_More than 1'],
      dtype='object')

# Split old_appliants Data 1st Round

In [16]:
# using given random state to obtain the same split
# You should choose 70% to create the Estimation set and 30% to create the Validation set. 
# Set the Random Seed to 1 if you're using Alteryx.

X = old_applicants_dummies.loc[:, ~old_applicants_dummies.columns.isin(['Credit-Application-Result_Non-Creditworthy'])].copy()
y = y_dummy = old_applicants_dummies['Credit-Application-Result_Non-Creditworthy'].copy()


X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 1, train_size = 0.7)



In [17]:
# impute missing information w/ median value of training set. Age-years is the only feature with missing data.
training_median = X_train.median()
X_train = X_train.fillna(training_median)
X_test = X_test.fillna(training_median)

print(training_median)

Duration-of-Credit-Month                             18.0
Credit-Amount                                      2176.0
Instalment-per-cent                                   3.0
Most-valuable-available-asset                         2.0
Age-years                                            33.0
Type-of-apartment                                     2.0
Occupation                                            1.0
No-of-dependents                                      1.0
Telephone                                             1.0
Foreign-Worker                                        1.0
Account-Balance_Some Balance                          0.0
Payment-Status-of-Previous-Credit_Paid Up             1.0
Payment-Status-of-Previous-Credit_Some Problems       0.0
Purpose_New car                                       0.0
Purpose_Other                                         0.0
Purpose_Used car                                      0.0
Value-Savings-Stocks_None                             1.0
Value-Savings-

In [18]:
X_train.isnull().sum()

Duration-of-Credit-Month                           0
Credit-Amount                                      0
Instalment-per-cent                                0
Most-valuable-available-asset                      0
Age-years                                          0
Type-of-apartment                                  0
Occupation                                         0
No-of-dependents                                   0
Telephone                                          0
Foreign-Worker                                     0
Account-Balance_Some Balance                       0
Payment-Status-of-Previous-Credit_Paid Up          0
Payment-Status-of-Previous-Credit_Some Problems    0
Purpose_New car                                    0
Purpose_Other                                      0
Purpose_Used car                                   0
Value-Savings-Stocks_None                          0
Value-Savings-Stocks_£100-£1000                    0
Length-of-current-employment_4-7 yrs          

In [19]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(350, 22)
(150, 22)
(350,)
(150,)


# Removing Features Based On Variance

In [20]:
from sklearn.feature_selection import VarianceThreshold
selector = VarianceThreshold(threshold = 0.0)
selector = selector.fit_transform(X_train)

In [21]:
X_train.var(axis=0).round(3)

Duration-of-Credit-Month                               146.560
Credit-Amount                                      7801794.006
Instalment-per-cent                                      1.248
Most-valuable-available-asset                            1.132
Age-years                                              131.136
Type-of-apartment                                        0.282
Occupation                                               0.000
No-of-dependents                                         0.139
Telephone                                                0.242
Foreign-Worker                                           0.036
Account-Balance_Some Balance                             0.251
Payment-Status-of-Previous-Credit_Paid Up                0.250
Payment-Status-of-Previous-Credit_Some Problems          0.044
Purpose_New car                                          0.076
Purpose_Other                                            0.025
Purpose_Used car                                       

In [22]:
# Notice Concurrent-Credits and Occupation have both been removed from the dataset. There are only 17 columns now.
selector.var(axis=0).round(3)

array([1.46141000e+02, 7.77950317e+06, 1.24500000e+00, 1.12900000e+00,
       1.30762000e+02, 2.81000000e-01, 1.38000000e-01, 2.41000000e-01,
       3.60000000e-02, 2.50000000e-01, 2.49000000e-01, 4.40000000e-02,
       7.60000000e-02, 2.50000000e-02, 1.42000000e-01, 2.40000000e-01,
       2.10000000e-01, 1.88000000e-01, 2.48000000e-01, 6.90000000e-02,
       2.35000000e-01])

In [23]:
# testing df_trfaining_corr dataset for low variance
X_train.var(axis=0)

Duration-of-Credit-Month                           1.465602e+02
Credit-Amount                                      7.801794e+06
Instalment-per-cent                                1.248465e+00
Most-valuable-available-asset                      1.132345e+00
Age-years                                          1.311363e+02
Type-of-apartment                                  2.818256e-01
Occupation                                         0.000000e+00
No-of-dependents                                   1.386492e-01
Telephone                                          2.418011e-01
Foreign-Worker                                     3.586574e-02
Account-Balance_Some Balance                       2.506836e-01
Payment-Status-of-Previous-Credit_Paid Up          2.495375e-01
Payment-Status-of-Previous-Credit_Some Problems    4.374949e-02
Purpose_New car                                    7.620958e-02
Purpose_Other                                      2.512485e-02
Purpose_Used car                        

In [24]:
# dropping columns with zero variance
# X_train = X_train.drop(columns = ['Occupation'])

In [25]:
# display dataset with columns dropped
# X_train.columns

Occupation is going to dropped from the dataset in the next section due to variance. Interestingly _Concurrent-Credits_ would have also been dropped due to having a variance of zero, but it was dropped from the dataset after running pd.get_dummies() function ealier. Therefore it did not appear in this section. I beleive this is due to it only having one value (i.e. it is dropped naturally because it only has one value _Other Banks/Depts_).

# Split old_appliants Data 2nd Round

In [26]:
# the first split allowed me to look at variance across all features that will be split.
# this second round is allowing me to re-copy the original dataset and drop the attributes I do not want
# in this case I am dropping Occupation due to variance
# there is probably a more elegant way to do this, but I have not figured that out yet

X = old_applicants_dummies.loc[:, ~old_applicants_dummies.columns.isin(['Credit-Application-Result_Non-Creditworthy', 'Occupation'])].copy()
y = y_dummy = old_applicants_dummies['Credit-Application-Result_Non-Creditworthy'].copy()

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 1, train_size = 0.7)


In [27]:
# impute missing information w/ median value of training set. Age-years is the only feature with missing data.
training_median = X_train.median()
X_train = X_train.fillna(training_median)
X_test = X_test.fillna(training_median)

print(training_median)

Duration-of-Credit-Month                             18.0
Credit-Amount                                      2176.0
Instalment-per-cent                                   3.0
Most-valuable-available-asset                         2.0
Age-years                                            33.0
Type-of-apartment                                     2.0
No-of-dependents                                      1.0
Telephone                                             1.0
Foreign-Worker                                        1.0
Account-Balance_Some Balance                          0.0
Payment-Status-of-Previous-Credit_Paid Up             1.0
Payment-Status-of-Previous-Credit_Some Problems       0.0
Purpose_New car                                       0.0
Purpose_Other                                         0.0
Purpose_Used car                                      0.0
Value-Savings-Stocks_None                             1.0
Value-Savings-Stocks_£100-£1000                       0.0
Length-of-curr

In [28]:
X_train.isnull().sum()

Duration-of-Credit-Month                           0
Credit-Amount                                      0
Instalment-per-cent                                0
Most-valuable-available-asset                      0
Age-years                                          0
Type-of-apartment                                  0
No-of-dependents                                   0
Telephone                                          0
Foreign-Worker                                     0
Account-Balance_Some Balance                       0
Payment-Status-of-Previous-Credit_Paid Up          0
Payment-Status-of-Previous-Credit_Some Problems    0
Purpose_New car                                    0
Purpose_Other                                      0
Purpose_Used car                                   0
Value-Savings-Stocks_None                          0
Value-Savings-Stocks_£100-£1000                    0
Length-of-current-employment_4-7 yrs               0
Length-of-current-employment_< 1yr            

In [29]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(350, 21)
(150, 21)
(350,)
(150,)


# <span style="color:red">Visualizations</span>

# <span style="color:red">Correlation</span>

# <span style="color:red">P-Value<span>

# Logistic Regression

In [30]:
# create classifier as empty box
LRC = LogisticRegression()

In [31]:
# my classifer doesn't now anything about being credit worthy or not
# need to use fit learning algorithm to train the classifier
LRC.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [32]:
# let's take the classifier for a spin and make some predictions
LRC.predict(X_test)

array([0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=uint8)

In [33]:
# Score it on the test set
# Using the default classification accuracy score
LRC.score(X_test, y_test)

0.76

In [34]:
# get same score as above
# from sklearn.metrics import classification_report
# classification_report(y_test, LRC.predict(X_test))


# Decision Tree

In [35]:
# create classifer... right now it's just an empty box
dectree = tree.DecisionTreeClassifier()


In [36]:
# my classifier doesn't know anything about being credit worthy or not yet
# to train the classifier, I need a learning algorithm
# if a classifier is a box of rules, then the learning algorithm is the procedure that creates the rules
# fit is my classifier, think of it as a synonym for find patterns in data
dectree.fit(X_train, y_train)


DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [37]:
# now that the classifier has been trained let's take it for a spin
dectree.predict(X_test)


array([0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0], dtype=uint8)

In [38]:
# Score it on the test set
# Using the default classification accuracy score
dectree.score(X_test, y_test)


0.6666666666666666

In [39]:
creditworthy_results = ["Creditworthy", "Non-Creditworthy"]

In [40]:
list(X.columns)

['Duration-of-Credit-Month',
 'Credit-Amount',
 'Instalment-per-cent',
 'Most-valuable-available-asset',
 'Age-years',
 'Type-of-apartment',
 'No-of-dependents',
 'Telephone',
 'Foreign-Worker',
 'Account-Balance_Some Balance',
 'Payment-Status-of-Previous-Credit_Paid Up',
 'Payment-Status-of-Previous-Credit_Some Problems',
 'Purpose_New car',
 'Purpose_Other',
 'Purpose_Used car',
 'Value-Savings-Stocks_None',
 'Value-Savings-Stocks_£100-£1000',
 'Length-of-current-employment_4-7 yrs',
 'Length-of-current-employment_< 1yr',
 'Guarantors_Yes',
 'No-of-Credits-at-this-Bank_More than 1']

In [41]:
# This output can be visualized
import pydot
from sklearn.externals.six import StringIO

dotfile = StringIO()
tree.export_graphviz(dectree,
                     out_file=dotfile,
                     feature_names=list(X.columns),
                     class_names=creditworthy_results,
                     filled=True,
                     rounded=True,
                     impurity=False
                    )
(graph,)=pydot.graph_from_dot_data(dotfile.getvalue())
graph.write_png("./assets/dtree.png")
graph.write_pdf("./assets/dtree.pdf")

This is a diagram of the visuals I saved to my repository. The diagram provides a visual representation of the decision tree. The nice thing about decision trees is that they are interpretable, and the tree diagram works the same way it would in code (i.e. you know exactly why the classifier makes a decision).

![Displaying the visual I just created:](./assets/dtree.png)


# Boosted Tree (aka Gradient Boosting)


In [42]:
# create classifier... this is an empty box
gbc = GradientBoostingClassifier()

In [43]:
# my classifer doesn't now anything about being credit worthy or not
# use fit algorithm to train the gbc classifier
gbc.fit(X_train, y_train)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=100,
              presort='auto', random_state=None, subsample=1.0, verbose=0,
              warm_start=False)

In [44]:
# take your classifier for a spin, make some predictions
gbc.predict(X_test)

array([0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=uint8)

In [45]:
# Score it on the test set
# Using the default classification accuracy score
gbc.score(X_test, y_test)

0.7266666666666667

# Forest Model (aka Random Forest Model)

In [46]:
# classifier is empty box
rf = RandomForestClassifier()

In [47]:
# classifier doesn't know anything about being credit worthy
# so need to use fit algorithm to train the classifier
# fit(features, target)
rf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [48]:
# take the trained classifier for a spin and make some predictions
rf.predict(X_test)

array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=uint8)

In [49]:
# Score it on the test set
# Using the default classification accuracy score
rf.score(X_test, y_test)

0.7466666666666667

In [50]:
BREAK

NameError: name 'BREAK' is not defined

# Test New Data Set

* get dummies for new dataset
* drop Occupation and Concurrent from data set. If needed, you might just select features based on previous section...
* but if you need to drop columns then you can using the code in next cell

In [ ]:

# X = old_applicants_dummies.loc[:, ~old_applicants_dummies.columns.isin(['Credit-Application-Result_Non-Creditworthy', 'Occupation'])].copy()
# y = y_dummy = old_applicants_dummies['Credit-Application-Result_Non-Creditworthy'].copy()


# Correlation

Note:

When when two features are highly correlated they are linearly dependent, and will have the same or similar effect on the dependent variable (target variable???). Therefore, when two features have high correlation, one of the two features can be dropped.

??? using the funciton np.random.seed() allowed me to make random numbers predictable and replicable ???
??? we randomize the order of data to affect how the model is trained (i.e. don't want order to affect how we learn)... order of the data  is not a part in determining if an applicant is creditworthy or not.... 
https://www.youtube.com/watch?v=nKW8Ndu7Mjw


??? do i need to create dummy variables first, before doing the following steps... to find out start small with 2 columns and see if you get the same values???

labeled the categorical variable using scikit-learn's LabelEncoder() class to transform non-numberical data to numerical data types.

Then calcualted each column's variance and removed items with low variance. In this case, I removed columns Concurrent-Credits and Occupation for both having a variance of zero; meaning they had the same value accross the entire field. 

then generated correlation matrix with old_applicants.corr(), which I passed to a correlation heat-map using seaborn. 

I then wrote a script that allowed me to compare the correlation between features. If two features had a correlation higher than 0.7, then one of the features was removed. My features were selected once the dataset only had the columns with correlation less than 0.7.



## Selecting Features Based On Correlation

Note:

[scikit-learn VerianceThreshold](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.VarianceThreshold.html) can be used to remove features with low variance. The default variance is set at 0.0. In this case Concurrent-Credits and Occupation variability are equal to zero. They will be removed from the dataframe.

In [ ]:
corr = old_applicants_dummies.loc[:, ~old_applicants_dummies.columns.isin(['Concurrent-Credits','Occupation'])].copy()
corr = corr.corr()

In [ ]:
corr.shape

In [ ]:
corr

In [ ]:
# GENERATE THE HEATMAP
sns.heatmap(corr)

The heatmap function automatically chooses the most correlated features, and I want to focus on features correlations to the target variable. Therfore I am assuming the more coreelatated a feature is to a target, the more useful it will be in my model. Also, correlation coefficient can be used to determine similar features (i.e. redundancies). To reduce overfitting my model, I will be using correlation to spot overfitting and remove redundant features.
https://hub.packtpub.com/selecting-statistical-based-features-in-machine-learning-application/ 
<span style="color:red">_this covers chapter 5... look at chapter 4 for additional details_</span>

In [ ]:
# isolating correlations between features and the target variable
corr['Credit-Application-Result']

I'm ignoring the correlation for _Credit-Application-Result_ since it is the target variable. Then looking at all of the features that have correlations close to -1 or +1; since these are the features I am assuming will be useful. 

In [ ]:
# using pandas to filter features that have ±0.2 correlations. 
corr['Credit-Application-Result'].abs() > 0.7

Every False in the preceding pandas Series represents a feature that has a correlation value between -.7 and .7 inclusive, while True values correspond to features with preceding correlation values .7 or less than -0.7.

Each _False_ in the series above represents a feature correlation between ±0.7 (inclusive). While, _True_ represents feature correlation values less than -0.7, or greater than 0.7. This can be plugged into the previous filter as a mask. I will assign it to a variable that represents highly correlated features and will drop the target variable. This should leave me with a list of highly correlated features.

In [ ]:
# plugging mask into previous pandas filter
highly_correlated_features = corr.columns[corr['Credit-Application-Result'].abs() > 0.7]
highly_correlated_features = highly_correlated_features.drop('Credit-Application-Result')
highly_correlated_features

# KNN

In [ ]:
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)
print(knn.score(X_test, y_test))

In [ ]:
NOW YOU CAN RUN CORR ON THE DATA

In [ ]:
BREAK

# Logistic Regression

In [ ]:
# old_applicants_dummies.drop('Credit-Application-Result_Non-Creditworthy', axis=1)


In [ ]:
# X_dummy = old_applicants_dummies.loc[:, ~old_applicants_dummies.columns.isin(['Credit-Application-Result_Non-Creditworthy'])]


In [ ]:
# y_dummy = old_applicants_dummies['Credit-Application-Result_Non-Creditworthy']


In [ ]:
# X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.7, random_state = 1)


In [ ]:
# X_train.isnull().sum()

In [ ]:
# X_train['Age-years'].head(10)

In [ ]:
# train_dummy_median = X_train_dummy.median()
# X_train_dummy = X_train_dummy.fillna(train_dummy_median)
# X_test_dummy = X_test_dummy.fillna(train_dummy_median)


In [ ]:
# X_train_dummy['Age-years'].head(10)

In [ ]:
# from sklearn.linear_model import LogisticRegression

In [ ]:
logmodel = LogisticRegression()

In [ ]:
logmodel.fit(X_train, y_train)

In [ ]:
predictions = logmodel.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
classification_report(y_test, predictions)

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy_score(y_test, predictions)

# Feature Selection

In [ ]:
# import warnings
# warnings.filterwarnings("ignore")
# from sklearn.model_selection import train_test_split
# from sklearn.svm import SVC
# from sklearn.metrics import confusion_matrix
# np.random.seed(1)

# <span style="color:red">START: Change variable name old_applicants to something not referencing corr</span>

Pearson correlation is the defaut for pandas.corr() function. In this case I will use it to measure the relationship between columns. Pearson correlation requires each column be normall distributed (which I am not assuming). Although the recommended threshold for ignoring this threshold is set at greater than 500, I am going to ignore this requirement because the shape of this dataset is right at 500 (meaning my dataset is large).
https://hub.packtpub.com/selecting-statistical-based-features-in-machine-learning-application/

In [ ]:
df_correlation = old_applicants.copy()


In [ ]:
old_applicants.head()

In [ ]:
old_applicants.columns

In [ ]:
# encode the categorciacl variables
label_encoder = LabelEncoder()
# encode the categorciacl variables
label_encoder = LabelEncoder()
df_correlation.iloc[:,0] = label_encoder.fit_transform(df_correlation.iloc[:,0]).astype('float64')
df_correlation.iloc[:,1] = label_encoder.fit_transform(df_correlation.iloc[:,1]).astype('float64')
df_correlation.iloc[:,2] = label_encoder.fit_transform(df_correlation.iloc[:,2])
df_correlation.iloc[:,3] = label_encoder.fit_transform(df_correlation.iloc[:,3]).astype('float64')
df_correlation.iloc[:,4] = label_encoder.fit_transform(df_correlation.iloc[:,4]).astype('float64')
df_correlation.iloc[:,5] = label_encoder.fit_transform(df_correlation.iloc[:,5])
df_correlation.iloc[:,6] = label_encoder.fit_transform(df_correlation.iloc[:,6]).astype('float64')
df_correlation.iloc[:,7] = label_encoder.fit_transform(df_correlation.iloc[:,7]).astype('float64')
df_correlation.iloc[:,8] = label_encoder.fit_transform(df_correlation.iloc[:,8])
df_correlation.iloc[:,9] = label_encoder.fit_transform(df_correlation.iloc[:,9]).astype('float64')
df_correlation.iloc[:,10] = label_encoder.fit_transform(df_correlation.iloc[:,10])
df_correlation.iloc[:,11] = label_encoder.fit_transform(df_correlation.iloc[:,11])
df_correlation.iloc[:,12] = label_encoder.fit_transform(df_correlation.iloc[:,12])
df_correlation.iloc[:,13] = label_encoder.fit_transform(df_correlation.iloc[:,13]).astype('float64')
df_correlation.iloc[:,14] = label_encoder.fit_transform(df_correlation.iloc[:,14])
df_correlation.iloc[:,15] = label_encoder.fit_transform(df_correlation.iloc[:,15]).astype('float64')
df_correlation.iloc[:,16] = label_encoder.fit_transform(df_correlation.iloc[:,16])
df_correlation.iloc[:,17] = label_encoder.fit_transform(df_correlation.iloc[:,17])
df_correlation.iloc[:,18] = label_encoder.fit_transform(df_correlation.iloc[:,18])








# # encode the categorciacl variables
# label_encoder = LabelEncoder()
# # encode the categorciacl variables
# label_encoder = LabelEncoder()
# old_applicants.iloc[:,0] = label_encoder.fit_transform(old_applicants.iloc[:,0]).astype('float64')
# old_applicants.iloc[:,1] = label_encoder.fit_transform(old_applicants.iloc[:,1]).astype('float64')
# old_applicants.iloc[:,2] = label_encoder.fit_transform(old_applicants.iloc[:,2])
# old_applicants.iloc[:,3] = label_encoder.fit_transform(old_applicants.iloc[:,3]).astype('float64')
# old_applicants.iloc[:,4] = label_encoder.fit_transform(old_applicants.iloc[:,4]).astype('float64')
# old_applicants.iloc[:,5] = label_encoder.fit_transform(old_applicants.iloc[:,5])
# old_applicants.iloc[:,6] = label_encoder.fit_transform(old_applicants.iloc[:,6]).astype('float64')
# old_applicants.iloc[:,7] = label_encoder.fit_transform(old_applicants.iloc[:,7]).astype('float64')
# old_applicants.iloc[:,8] = label_encoder.fit_transform(old_applicants.iloc[:,8])
# old_applicants.iloc[:,9] = label_encoder.fit_transform(old_applicants.iloc[:,9]).astype('float64')
# old_applicants.iloc[:,10] = label_encoder.fit_transform(old_applicants.iloc[:,10])
# old_applicants.iloc[:,11] = label_encoder.fit_transform(old_applicants.iloc[:,11])
# old_applicants.iloc[:,12] = label_encoder.fit_transform(old_applicants.iloc[:,12])
# old_applicants.iloc[:,13] = label_encoder.fit_transform(old_applicants.iloc[:,13]).astype('float64')
# old_applicants.iloc[:,14] = label_encoder.fit_transform(old_applicants.iloc[:,14])
# old_applicants.iloc[:,15] = label_encoder.fit_transform(old_applicants.iloc[:,15]).astype('float64')
# old_applicants.iloc[:,16] = label_encoder.fit_transform(old_applicants.iloc[:,16])
# old_applicants.iloc[:,17] = label_encoder.fit_transform(old_applicants.iloc[:,17])
# old_applicants.iloc[:,18] = label_encoder.fit_transform(old_applicants.iloc[:,18])


In [ ]:
# Label Encoder somehow removes my null Age-years values
df_correlation.isnull().sum()

In [ ]:
BREAK

In [ ]:
# X = old_applicants[['Age-years', 'Account-Balance', 'Duration-of-Credit-Month','Payment-Status-of-Previous-Credit']].copy()
# y = old_applicants['Credit-Application-Result'].copy()

In [ ]:
## This code works to impute median value on age, but I want to paly around with Imputer transformer instead

# # median age to impute missing values with
# train_imputeAge = old_applicants['Age-years'].median()
# train_imputeAge

# # fill missing age values with imputed values
# train['Age-years'].fillna(train_imputeAge, inplace=True)

# # calculate average of ages after imputing the mean value
# train['Age-years'].mean()

# <span style="color:red">Red Flag... imputing values before splitting is improper approach</span>
Note that we aren't implementing any grid searching here, just a plain fit. We see that our model boasts a 66% accuracy rate (not great, but that's not the point). The important thing to note here is that both the training and the testing set of X were imputed using the mean of the entire X matrix. This is in direct violation of a core tenet of the machine learning procedure. We cannot assume that we know the mean of the entire dataset when predicting the test set's response values. Simply put, our KNN model is using information gained from the testing set to fit to the training set. This is a big red flag. These values will have to be imputed after the split. However, my data might already be split into training and testing data.
https://subscription.packtpub.com/book/big_data_and_business_intelligence/9781787287600/3/ch03lvl1sec28/dealing-with-missing-values-in-a-dataset?_ga=2.100741634.2075634726.1552106478-132390188.1552018386

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
# import Imputer module from the scikit-learn preprocessing class
from sklearn.preprocessing import Imputer

In [ ]:
# use parameter strategy, and set it to median to define how you want to impute values into your dataset
imputer = Imputer(strategy='median')

In [ ]:
# call the fit_transform method to create a new object
tain_imputed = imputer.fit_transform(old_applicants[['Age-years']])

In [ ]:
# Imputer output is not a pandas DataFrame, but a NumPy array
type(tain_imputed)

In [ ]:
# mean of Age-years before imputing median values on missing records
old_applicants['Age-years'].mean()

In [ ]:
# replace Age-years of original dataset with imputed NumPy array... previous records with NaN value will now contain the median 33
traold_applicantsin['Age-years'] = tain_imputed

In [ ]:
type(old_applicants['Age-years'])

In [ ]:
# average Age-years has changed to reflect the mean with imputed values included
old_applicants['Age-years'].mean()

# <span style="color:red">BREAK CORRELATION</span>

In [ ]:
BREAK

Note: 

Notice there are still 17 columns, meaning no features have correlation of 0.7 or higher. Therefore no features will be removed from the dataset based on correlation value.

In [ ]:
columns = np.full((corr.shape[0],), True, dtype=bool)
for i in range(corr.shape[0]):
    for j in range(i+1, corr.shape[0]):
        if corr.iloc[i,j] >= 0.7:
            if columns[j]:
                columns[j] = False

# print(old_applicants.shape)
# print(columns)
# print(old_applicants.columns)

selected_columns = old_applicants.columns[columns]
data = old_applicants[selected_columns]

######
######
######

selected_columns = selected_columns[1:].values
import statsmodels.formula.api as sm
def backwardElimination(x, Y, sl, columns):
    numVars = len(x[0])
#     print(numVars)
    for i in range(0, numVars):
        print(x.shape)
        print(Y.shape)
        print(i)
        model = sm.OLS(Y, x[:,i])
        fit = model.fit()
        print(fit.pvalues)
        print(pearsonr(x[:,i],Y))
        maxVar = max(fit.pvalues).astype(float)
#         print(maxVar)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (fit.pvalues[j].astype(float) == maxVar):
                    x = np.delete(x, j, 1)
                    columns = np.delete(columns, j)
                    
    fit.summary()
#     print(x)
#     print(columns)
    return x, columns

SL = 0.05 # eliminating features which have p > SL... p-value gives us the probability of finding the observed or a more extreme result when the null hypothesis is true. So when the p-value increases with the removal or addition of a feature, the probability of obtaining the observed results also increases.
data_modeled, selected_columns = backwardElimination(data.iloc[:,1:].values, data.iloc[:,0].values, SL, selected_columns)



In [ ]:
pearsonr(data['Account-Balance'], data['Credit-Application-Result'])
# data.head()

In [ ]:
# BREAK

In [ ]:
# OLS: https://www.youtube.com/watch?v=L_h7XFUGWAk
selected_columns = selected_columns[1:].values
import statsmodels.formula.api as sm
def backwardElimination(x, Y, sl, columns):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(Y, x).fit()
        print(regressor_OLS.pvalues)        
        maxVar = max(regressor_OLS.pvalues).astype(float)
        print(maxVar)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    x = np.delete(x, j, 1)
                    columns = np.delete(columns, j)
                    
    regressor_OLS.summary()
#     print(x)
#     print(columns)
    return x, columns

SL = 0.05 # eliminating features which have p > SL... p-value gives us the probability of finding the observed or a more extreme result when the null hypothesis is true. So when the p-value increases with the removal or addition of a feature, the probability of obtaining the observed results also increases.
data_modeled, selected_columns = backwardElimination(data.iloc[:,1:].values, data.iloc[:,0].values, SL, selected_columns)


In [ ]:
for i in data:
    print(i)
    print("p-val:",pearsonr(data[i], result['diagnosis'])[0])
    print("Corr: ",pearsonr(data[i], result['diagnosis'])[1])
    print()


In [ ]:
# this is target value
result = pd.DataFrame()
result['diagnosis'] = data.iloc[:,0]
result['diagnosis'].head()

In [ ]:
# this dataframe of downselected features
data = pd.DataFrame(data = data_modeled, columns = selected_columns)
data.head()

Note:

The data set has been downselected to 6 features:
* _Account-Balance_
* _Duration-of-Credit-Month_
* _Payment-Status-of-Previous-Credit_
* _Length-of-current-employment_
* _Instalment-per-cent_
* _Most-valuable-available-asset_

## Visualizing the Data

In [ ]:
# sns.countplot(data=data, x=result['diagnosis'], hue='Account-Balance')
# sns.distplot(data=data, x=result['diagnosis'], hue='Account-Balance')
# sns.distplot(data['Account-Balance'][result['diagnosis']==0, color='b', label = 'Crediworthy', kde=False)
# sns.distplot(data['Account-Balance'], color='r', label = 'Non-Crediworthy', kde=False)

# <span style="color: red">BREAK</span>

fix graphs... used original table explain that you are using them out of simplicity... then go into stats models

In [ ]:
# old_applicants.columns

In [ ]:
# analyze the results of credit applicatons

sns.countplot(data=old_applicants, x="Credit-Application-Result")

In [ ]:
sns.countplot(data=old_applicants, x='Credit-Application-Result', hue='Account-Balance')

In [ ]:
sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Account-Balance")

In [ ]:
sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Duration-of-Credit-Month")
# ax.legend(loc=4)
plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ).set_title("Duration-of-Credit-Month")


In [ ]:
sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Payment-Status-of-Previous-Credit")


In [ ]:
sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Length-of-current-employment")

In [ ]:
sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Instalment-per-cent")

In [ ]:
sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Most-valuable-available-asset")

In [ ]:
# Account-Balance
# Duration-of-Credit-Month
# Payment-Status-of-Previous-Credit
# Length-of-current-employment
# Instalment-per-cent
# Most-valuable-available-asset

In [ ]:
# sns.countplot(data=old_applicants, x='Credit-Application-Result', hue='Account-Balance')
# sns.countplot(data=data, x=result['diagnosis'], hue='Account-Balance')

Taking a look at relationship between the target variable (i.e. _Credit-Application-Result_) and all other features in the training set.

There are more creditworthy applicants (≈350) vs non-creditworth applicants (≈150)

Out of all the applicantion results who was creditworth and who was not credithworthy in regards to Account-Balance. More people were deemed Creditworth if they had some balance.

Tough to determine any relationship regarding Duration-of-Credit-Month.

Regarding payment-Status-of-Previous-Credit, More people were marked as creditworthy if they were  

In [ ]:
dummy variables

# response variable
"Credit-Application-Result_Non-Creditworthy" 

# target variables
"Duration-of-Credit-Month"
"Credit-Amount"
"Instalment-per-cent"
"Most-valuable-available-asset"
"Age-years"
"Type-of-apartment"
"No-of-dependents"
"Telephone"
"Foreign-Worker"
"Account-Balance_Some Balance"
"Payment-Status-of-Previous-Credit_Paid Up"
"Payment-Status-of-Previous-Credit_Some Problems"
"Purpose_New car"
"Purpose_Other"
"Purpose_Used car"
"Value-Savings-Stocks_None"
"Value-Savings-Stocks_£100-£1000"
"Length-of-current-employment_4-7 yrs"
"Length-of-current-employment_< 1yr"
"Guarantors_Yes"
"No-of-Credits-at-this-Bank_More than 1"



In [ ]:
# sns.countplot(data=train, x="Credit-Application-Result", hue="Purpose")
# More people were marked as creditworthy if the purpose was home related

In [ ]:
# sns.countplot(data=train, x="Credit-Application-Result", hue="Credit-Amount")


In [ ]:
# bins = 5000
# x = old_applicants["Credit-Amount"][old_applicants["Credit-Application-Result"] == "Creditworthy"] 
# y = old_applicants["Credit-Amount"][old_applicants["Credit-Application-Result"] == "Non-Creditworthy"] 

# plt.hist(x, bins, alpha=0.5, label='x')
# plt.hist(y, bins, alpha=0.5, label='y')
# plt.legend(loc='upper right')
# plt.show()



# the histogram of the data
# plt.hist(train["Credit-Amount"], bins=1000)
# plt.show()


In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Value-Savings-Stocks")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Guarantors")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Age-years")

I don't know how _Age_ didn't make the list. Will need to investigate further.

In [ ]:
# https://markhneedham.com/blog/2013/01/27/pythonnumpy-selecting-specific-column-in-2d-array/
# To select specific column use the following format... basically stating select all rows from pecific index
# zeroth column is Credit-Application-Result
selector[:,0]

# Account Balance has index 1
selector[:,1]


In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Concurrent-Credits")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Type-of-apartment")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="No-of-Credits-at-this-Bank")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Occupation")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="No-of-dependents")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Telephone")

In [ ]:
# sns.countplot(data=old_applicants, x="Credit-Application-Result", hue="Foreign-Worker")

In [ ]:
# fig = plt.figure(figsize = (20, 25))
# j = 0
# for i in data.columns:
#     plt.subplot(6, 4, j+1)
#     j += 1
#     sns.distplot(data[i][result['diagnosis']==0], color='g', label = 'Creditworthy', kde=False)
#     sns.distplot(data[i][result['diagnosis']==1], color='r', label = 'Non-Crediworthy', kde=False)
#     plt.legend(loc='best')
# fig.suptitle('Default Risk Data Analysis')
# fig.tight_layout()
# fig.subplots_adjust(top=0.95)
# plt.savefig("./assets/down selected features.png", psi=300)
# plt.show()

In [ ]:
# old_applicants.head()

In [ ]:
# selector[0]

In [ ]:
# sns.countplot(data=selector[:,0:2], x=selector[:,0], hue=selector[:,1])
# sns.distplot(data[i][result['diagnosis']==0], color='g', label = 'Creditworthy', kde=False)
# sns.distplot(data[0][result['diagnosis']==1], color='r', label = 'Non-Crediworthy', kde=False)


In [ ]:
# sns.countplot(data=selector[:,0:2], x=selector[:,0], hue=selector[:,1])
# sns.countplot(data=old_applicants, x='Credit-Application-Result', hue='Account-Balance')


In [ ]:
# sns.countplot(data=selector[:,0:2], x=selector[:,0], hue=selector[:,1])
# sns.countplot(data=old_applicants, x='Account-Balance', hue='Credit-Application-Result')


In [ ]:
# sns.countplot(data=train, x="Credit-Application-Result", hue="Account-Balance")

In [ ]:
# old_applicants["Account-Balance"].unique()

In [ ]:
# data["Account-Balance"].unique()

In [ ]:
# data["Length-of-current-employment"].unique()

In [ ]:
# test for variance across dataset
# data.var(axis=0)

In [ ]:
# old_applicants.var(axis=0)

In [ ]:
# BREAK

In [ ]:
# Look into correlations for the following fields:
# possPredictors = ["Account-Balance",
#                   "Payment-Status-of-Previous-Credit",
#                   "Purpose",
#                   "Credit-Amount",
#                   "Value-Savings-Stocks",
#                   "Length-of-current-employment",
#                   "Instalment-per-cent",
#                   "Guarantors",
#                   "Most-valuable-available-asset",
#                   "Age-years",
#                   "Concurrent-Credits",
#                   "Type-of-apartment",
#                   "No-of-Credits-at-this-Bank",
#                   "Occupation",
#                   "No-of-dependents",
#                   "Telephone",
#                   "Foreign-Worker"
#                 ]


In [ ]:
# old_applicants["Credit-Application-Result"]=old_applicants["Credit-Application-Result"].astype('category').cat.codes
# old_applicants["Account-Balance"]=old_applicants["Account-Balance"].astype('category').cat.codes
# old_applicants["Credit-Application-Result"].dtype


In [ ]:
# first create dummies, and drop unwanted va

# https://code.i-harness.com/en/q/186322a
# https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.pearsonr.html#scipy.stats.pearsonr
# pearson(x,y)... or predictor and target
# Where the first value in the tuple is the correlation value, and second is the p-value.
# creditworthy = np.where(old_applicants["Credit-Application-Result"]=="Creditworthy", 1, 0)

from scipy.stats import pearsonr
for i in data:
    print(i)
    print("p-val:",pearsonr(data[i], result['diagnosis'])[0])
    print("Corr: ",pearsonr(data[i], result['diagnosis'])[1])
    print()


In [ ]:
# from scipy.stats import spearmanr
# for i in data:
#     print(i)
#     print("p-val:",spearmanr(data[i], result['diagnosis'])[0])
#     print("Corr: ",spearmanr(data[i], result['diagnosis'])[1])
#     print()



In [ ]:
BREAK
# pearsonr(old_applicants["Credit-Amount"], creditworthy)

In [ ]:
old_applicants.info()
# old_applicants.head()

In [ ]:
# data_cols = ["Credit-Application-Result",
#              "Account-Balance",
#              "Payment-Status-of-Previous-Credit",
#              "Purpose",
#              "Value-Savings-Stocks",
#              "Guarantors",
#              "Concurrent-Credits",
#              "No-of-Credits-at-this-Bank"]

# label_encoder = LabelEncoder()
# for i in data_cols:
#     data.loc[:,i] = label_encoder.fit_transform(data.loc[:,i]).astype('float64')

In [ ]:
# dummy columns
dummy_cols = ["Account-Balance", "Purpose"]

# Purpose_Home Related, Purpose_New car, Purpose_Other, Purpose_Used car
# Account-Balance_No Account, Account-Balance_Some Balance

old_applicants = pd.get_dummies(old_applicants, columns=dummy_cols, drop_first=True)

In [ ]:
old_applicants.iloc[0:5,0:5]

In [ ]:
old_applicants["Credit-Application-Result"].unique()

In [ ]:
pearsonr(old_applicants["Account-Balance"], old_applicants["Credit-Application-Result"])

In [ ]:
break


In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
import warnings
warnings.filterwarnings("ignore")
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix

In [ ]:
data = old_applicants.copy()
# data = data.iloc[:,1:-1]

In [ ]:
data.head()

In [ ]:
data_cols = ["Credit-Application-Result",
             "Account-Balance",
             "Payment-Status-of-Previous-Credit",
             "Purpose",
             "Value-Savings-Stocks",
             "Guarantors",
             "Concurrent-Credits",
             "No-of-Credits-at-this-Bank"]

label_encoder = LabelEncoder()
for i in data_cols:
    data.loc[:,i] = label_encoder.fit_transform(data.loc[:,i]).astype('float64')

In [ ]:
corr = data.corr()


In [ ]:
sns.heatmap(corr)

In [ ]:
# # COLORMAPS

# # use cmap to set the chart colors
# cmap = sns.choose_colorbrewer_palette(data_type = "d") # set data_type to d, q, or s

In [ ]:
# CORRELATION MATRIX
# def correlationMatrix(dataset,suffix):
#     corr = dataset.corr(method='pearson')
    
#     # Generate a mask for the upper triangle
#     mask = np.zeros_like(corr, dtype=np.bool)
#     mask[np.triu_indices_from(mask)] = True

#     # Set up the matplotlib figure
# #     fig, ax = plt.subplots(figsize=(10, 6))
#     fig, ax = plt.subplots()

#     # Draw the heatmap with the mask and correct aspect ratio
#     sns.heatmap(corr, annot=True, fmt='.4f',
#                 cmap=cmap, cbar=True, ax=ax, mask = mask,
#                 square=False, linewidths=.5,cbar_kws={"shrink": .5})
#     ax.set_title('Heatmap', fontsize = 18)
#     ax.set_xlabel('Variable 1', fontsize = 15)
#     ax.set_ylabel('Variable 2', fontsize = 15)
#     ax.set_yticklabels(ax.get_yticklabels(), rotation="horizontal", fontsize = 10)
#     ax.set_xticklabels(ax.get_xticklabels(), fontsize = 10)
#     plt.tight_layout()
#     plt.savefig('./assets/pearsonCorrelation_{}'.format(suffix), dpi = 300, bbox_inches='tight', pad_inches=0.0)
#     plt.show()

In [ ]:
# correlationMatrix(corr,"old_applicants")

In [ ]:
# correlationMatrix(old_applicants,"old_applicants")

In [ ]:
# correlationMatrix(new_applicants,"new_applicants")

# LOGISTIC REGRESSION

In [ ]:
# dummy columns
dummy_cols = ["Account-Balance", "Purpose"]

In [ ]:
# Purpose_Home Related, Purpose_New car, Purpose_Other, Purpose_Used car
# Account-Balance_No Account, Account-Balance_Some Balance

old_applicants = pd.get_dummies(old_applicants, columns=dummy_cols, drop_first=True)

In [ ]:
old_applicants.columns

In [ ]:
new_applicants = pd.get_dummies(new_applicants, columns=dummy_cols)

In [ ]:
new_applicants.columns

# <span style = "color:red">I think X & y need to be split into train & test set... not predict on new_applicants dataset yet</span>

In [ ]:
# Dummies:'Account-Balance_No Account', 'Account-Balance_Some Balance', 
# Dummies:'Purpose_Home Related', 'Purpose_New car', 'Purpose_Used car'

# create list of feature names
feature_cols = ["Credit-Amount",
                "Account-Balance_Some Balance",
                "Purpose_New car",
                "Purpose_New car",
                "Purpose_Used car"
               ]

# store feature matrix in "X"... old_applicants data featres
X_train = old_applicants[feature_cols]

# store response vector in "y"... train data target
y_train = old_applicants["Credit-Application-Result"]

# predict new observations... new_applicants data
X_test = new_applicants[feature_cols]

In [ ]:
# import the class
from sklearn.linear_model import LogisticRegression

# instantiate the model (using the default parameters)
logreg = LogisticRegression()

# fit the model with data
logreg.fit(X_train, y_train)

# predict new targets based on new_applicants data
y_pred = logreg.predict(X_test)


In [ ]:
# looking at coefficient of the model, print the regression coefficient: https://www.youtube.com/watch?v=Eqv98w1ukZk
print(logreg.coef_)

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_train,y_pred))


In [ ]:
# correlationMatrix(old_applicants,"old_applicants")

In [ ]:
# correlationMatrix(old_applicants ,"test")